### PR TITLE
Fork from turtle activity and port to Gtk+ 3

### PR DESCRIPTION
**Pull request by yell0wfl4sh**
Fork of ```turtleart-activity``` from the present state, added required changes to make it ```Turtle Confusion``` activity.

Some changes I am not sure of:
-  NEWS file
-  Version metadata
There might be some other license related issue that I am not sure of.

@quozl @walterbender , I know there are way too many changes to be reviewed but I still request to have a look here and merge this pull request, thanks!

### CODE REVIEW

**Comment by walterbender** [comment]
It seems close. There seem to be a few things still broken (some of these were fixed in Quozl's patch from today):

/home/walter/Activities/activity-turtle-confusion/TurtleArt/tawindow.py:600: Warning: Source ID 1266 was not found when attempting to remove it
  self.activity._unfullscreen_button_timeout_id)

(See https://github.com/sugarlabs/turtleart-activity/commit/daa498e2d0c153c81b1dcf222ef9a28ac1c9ecdc)

Traceback (most recent call last):
  File "/home/walter/Activities/activity-turtle-confusion/TurtleArtActivity.py", line 491, in do_palette_buttons_cb
    self._help_button.set_current_palette(palette_names[i])
AttributeError: 'ToolButton' object has no attribute 'set_current_palette'

This one may be broken in Turtle Art as well.

Traceback (most recent call last):
  File "/home/walter/Activities/activity-turtle-confusion/TurtleArtActivity.py", line 599, in do_stop_cb
    self.stop_turtle_button.set_icon('hideshowoff')
AttributeError: 'ToolButton' object has no attribute 'set_icon'

should be set_icon_name (there may be other places where this is broken)

See https://github.com/sugarlabs/turtleart-activity/commit/547aa4756466abce45a2919634a2e0eb595d7b2e
https://github.com/sugarlabs/turtleart-activity/commit/b0f21da0db60e0b0f5e220f0291c70102f519904
and
https://github.com/sugarlabs/turtleart-activity/commit/4cb96a0ad3a7ad95b10249f54b6c2e81d7d5208c

**Comment by quozl** [comment]
Fixes https://github.com/sugarlabs/activity-turtle-confusion/issues/2

**Comment by yell0wfl4sh** [comment]
@walterbender @quozl added all commits from TurtleArt activity.

However, this error was logged doing simple operations on Turtle Confusion
```
Guessing python class from activity.info!
Traceback (most recent call last):
  File "/home/yash/testgsox/activity-turtle-confusion/TurtleArt/tawindow.py", line 3036, in __run_stack
    code = self.lc.generate_code(top, self.just_blocks())
  File "/home/yash/testgsox/activity-turtle-confusion/TurtleArt/talogo.py", line 278, in generate_code
    self.clear_value_blocks()
  File "/home/yash/testgsox/activity-turtle-confusion/TurtleArt/talogo.py", line 1045, in clear_value_blocks
    self.update_label_value(name)
  File "/home/yash/testgsox/activity-turtle-confusion/TurtleArt/talogo.py", line 1077, in update_label_value
    for block in self.value_blocks_to_update[name]:
KeyError: 'sound'
```
which is fixed in c1ec720f2861d75099fe63b6fe60420044c1e2aa .

please review

**Comment by quozl** [comment]
c1ec720 is not a correct fix, because we don't know why it happened, and `interactive_mode` is True anyway, so the change should have been ineffective but wasn't.

Unrelated issue in testing;

```
Traceback (most recent call last):
  File "/tmp/TurtleArtActivity.py", line 599, in do_stop_cb
    self.stop_turtle_button.set_icon('hideshowoff')
AttributeError: 'ToolButton' object has no attribute 'set_icon'
```

Should be `set_icon_name`; the code in this function is different between Turtle Art and Turtle Confusion.

**Comment by yell0wfl4sh** [comment]
Changed ```set_icon``` to ```set_icon_name``` in e91c34c2eb30ac54b719e82bc5b97cad8573f4af

**Comment by yell0wfl4sh** [comment]
If ```self.interactive_mode``` is true than ```self.lc.find_value_blocks()``` will not be executed. 
```
if not self.interactive_mode:
        self.lc.find_value_blocks()  # Are there blocks to update?
```
I think this line is necessary for this file to be executed properly.

**Comment by quozl** [comment]
@yashagrawal3, does this mean https://github.com/sugarlabs/turtleart-activity/commit/e822494254d233f0c9f9b080fa772c51786791d2#diff-deff5a37361d95d05f55434cd910e4b8L3028 is wrong and the `not` should be removed?

@walterbender said on IRC;
> and you do not want to find value blocks in non-interactive mode

and the code would seem to do the opposite.

If so, then your c1ec720 is not a correct fix, it should instead read;

```python
    if self.interactive_mode:
        self.lc.find_value_blocks()  # Are there blocks to update?
```

**Comment by yell0wfl4sh** [comment]
@quozl My bad, made changes in ebd529bcb4632fe0a6a8191f9587e1dab892d3e6.

**Comment by walterbender** [comment]
Still seeing these two errors:

```
Traceback (most recent call last):
  File "TurtleArtActivity.py", line 1859, in _create_store
    self._fill_challenges_list(store)
AttributeError: 'TurtleArtActivity' object has no attribute '_fill_challenges_list'
```
```
Traceback (most recent call last):
  File "TurtleArtActivity.py", line 491, in do_palette_buttons_cb
    self._help_button.set_current_palette(palette_names[i])
AttributeError: 'ToolButton' object has no attribute 'set_current_palette'
```

**Comment by quozl** [comment]
Agreed.  In ebd529b, `_fill_challenges_list` is not defined, and `_help_button` is not an instance of `HelpButton`.

**Comment by yell0wfl4sh** [comment]
Fixed both errors, please review again.

**Comment by quozl** [comment]
Reviewed ecd148b and 4196519.  Seems to be a mistake in two places;
* your addition of the call to `HelpButton` and assignment to `_help_button` is immediately discarded by the following line which calls `_add_button`,
* your removal of the call to `self._help_button.set_current_palette` in `do_palette_buttons_cb` would not then be needed,

Thanks for all the work on this; the repositories for Turtle Blocks and Turtle Confusion are now _much_ closer than before, and so I've carefully examined the differences.  For two reasons; (a) detect any unexplained changes during your merge, and (b) work toward merging the separate activities.

* `plugins/$NAME/$NAME.py` has functional change; makes blocks hidden; it would be better to add a keyword argument to each `__init__` so that the files can be identical between repositories.

* `plugins/camera_sensor/glive.py`, `instance.py`, `utils.py`; all three files seem unused, are GTK+ 2 and GStreamer 0.10 binding uses, please confirm and remove the files,

* there's a copy of `extrasoff.svg` and `extrason.svg` in each `plugins/*` directory; it would be better to have only one copy in a common place.  As the same problem affects Turtle Blocks for the `sensoroff.svg` and `sensoron.svg` files, it could be fixed later.

* `.gitignore` is different, easily resolved.

* `README.md` has unexpected differences; some of Turtle Blocks README.md might be added to Turtle Confusion, and vice versa.

* `setup.py` has an import missing from the Turtle Blocks side; if this code has lasted this long with a missing import, perhaps it is not needed, @walterbender?

* `TurtleArtActivity.py` could be changed to contain two activity classes, with a `TurtleConfusionActivity` class inheriting from `TurtleArtActivity`, and a corresponding change to _exec_ in `activity/activity.info`.

It looks possible to merge the _functionality_ of Turtle Confusion into Turtle Blocks.  If we do this, maintenance will be easier.

@yashagrawal3, are you interested in adding this to your project, or would you and your mentors prefer it be done outside your project?

**Comment by walterbender** [comment]
Would be great to combine them, Please keep in mind that there are several other versions of turtle confusion specific to different locals: Amazonas, Mexico, ...

**Comment by quozl** [comment]
Thanks.  Guess these are;
* Amazonas Tortuga [at walterbender](https://github.com/walterbender/AmazonasTortuga) or [at sugarlabs](https://github.com/sugarlabs/amazonas-activity) (same code but disconnected repositories), and;
* [Tortuga de Mexico](http://activities.sugarlabs.org/en-US/sugar/addon/4668) (not on GitHub),

I also see
* [Turtle Flags](https://github.com/sugarlabs/activity-turtle-flags),
* [Turtle Bots](https://github.com/nvazquez/Turtlebots), and;
* [Turtle Nutrition](http://activities.sugarlabs.org/en-US/sugar/addon/4742) (not on GitHub).

**Comment by walterbender** [comment]
Hmm. I will upload the repos for everything I can find (except Turtle Bots, which is maintained by a group in Uruguay -- I'll ask them to upload it to github).

**Comment by yell0wfl4sh** [comment]
@quozl  I will once talk to @chimosky about it and then tell you we will add this to our project.

File: TurtleArtActivity.py [hunk 1]
```diff
@@ -866,7 +866,6 @@ def _setup_toolbar(self):
         self.palette_toolbar_button = ToolbarButton(
             page=self._palette_toolbar, icon_name='palette')
 
-        self._help_button = HelpButton(self)
```
Comment (quozl) [inline TurtleArtActivity.py]: I don't understand why you have regressed the code at this point.  See https://github.com/sugarlabs/turtleart-activity/blob/master/TurtleArtActivity.py#L801

**Comment by yell0wfl4sh** [comment]
@quozl I have made changes that you suggested.

**Comment by quozl** [comment]
Thanks.  I've reviewed by;

* doing a rebase of Turtle Confusion (revert the two commits that diverged from Turtle Art, merge the Turtle Art changes since, then reapply the reverted commits and resolve conflicts),
* pushed to https://github.com/quozl/activity-turtle-confusion/commits/2018-06-01-rebase
* comparing your branch against mine, using `git diff`,

Some things you need to fix;

* `ObjectChooser` and `mime` modules are imported and not used,
* `self._local` and `self._custom_filepath` variables are defined and not used,
* `restore_state` and `_draw_cartoon` methods are missing,
* some of the `_logger.debug` calls got lost,
* `cancel_plugin_install` is incorrectly referenced,
* a Gtk.ScrolledWindow policy changed,
* ... and so on.

You can see the differences by using `git remote add` for my fork, fetch, then use `git diff quozl/2018-06-01-rebase`.  There are some bugs in my work too, and it is untested, but the differences will give you things to consider for investigation.

My guess is you've got several hours work ahead of you.  Have fun!

**Comment by yell0wfl4sh** [comment]
@quozl reviewed all difference between the two branch and made necessary changes, please see!

**Comment by quozl** [comment]
Why is `self._level` defined and not used?

**Comment by yell0wfl4sh** [comment]
it got left with ```self._custom_filepath``` ,  I have removed it now.

**Comment by quozl** [comment]
Thanks.  What is `self._custom_filepath` for?

* https://github.com/yashagrawal3/activity-turtle-confusion/blob/6ff8b5183951096b3569670d65b6d427407763f5/TurtleArtActivity.py#L112

* https://github.com/yashagrawal3/activity-turtle-confusion/blob/6ff8b5183951096b3569670d65b6d427407763f5/TurtleArtActivity.py#L542

* https://github.com/yashagrawal3/activity-turtle-confusion/blob/6ff8b5183951096b3569670d65b6d427407763f5/TurtleArtActivity.py#L1755

**Comment by yell0wfl4sh** [comment]
Since it is announced ```None``` and never changed later, I think it can be avoided but it is used in file later as mentioned, so avoided removing it.

**Comment by quozl** [comment]
Thanks.  In my branch, a call to `restore_cursor` is preceded by a comment "restore the cursor", and then a `_logger.debug()` with "restoring cursor", but in yours it says "setting watch cursor";

Compare mine vs yours;
* [mine](https://github.com/quozl/activity-turtle-confusion/blob/2018-06-01-rebase/TurtleArtActivity.py#L608),
* [yours](https://github.com/yashagrawal3/activity-turtle-confusion/blob/6ff8b5183951096b3569670d65b6d427407763f5/TurtleArtActivity.py#L611),

Compare with confusion vs original:
*  [this repository master](https://github.com/yashagrawal3/activity-turtle-confusion/blob/master/TurtleArtActivity.py#L552),
* [turtle art master](https://github.com/sugarlabs/turtleart-activity/blob/master/TurtleArtActivity.py#L570).

**Comment by yell0wfl4sh** [comment]
Changed to "restoring cursor", next line is evident that it is right.

**Comment by quozl** [comment]
Thanks.  You have a reference to `self._sample_window`, an undefined variable;

https://github.com/yashagrawal3/activity-turtle-confusion/blob/e48674ac65235ab30a16537c7a26d0b29a6d16e8/TurtleArtActivity.py#L1845

In my branch this is `self._challenge_window`.

**Comment by yell0wfl4sh** [comment]
Used `camera_sensor.py` becuase it was closer to `camera_sensor.py` in Turtle Activity.

**Comment by quozl** [comment]
Thanks.  Reviewed to 022c708.

Reminder; `plugins/camera_sensor/glive.py`, `instance.py`, `utils.py`; all three files seem unused, are GTK+ 2 and GStreamer 0.10 binding uses, please confirm and remove the files.

For your interest, another method I'm using for review is to compare two views of the change from Turtle Art to Turtle Confusion; [the most recent rebase by @walterbender](https://github.com/sugarlabs/activity-turtle-confusion/commit/2619b5235caec28d34da55a211e651e9e5eeac00) and the output of "git diff turtleart/master yashagrawal3/gtk3".  In my local repository, I've a remote added for the Turtle Art repository; because the two repositories share a common history.

**Comment by yell0wfl4sh** [comment]
`git grep gtk` shows that following files that are not ported to GTK3 and and after testing found that they were not being used by the activity:

- TurtleArt/sprites.py
- plugins/camera_sensor/glive.py
- plugins/camera_sensor/instance.py
- plugins/camera_sensor/utils.py
- pyexported/window_setup.py

I have removed all this files. Please verify these changes.

**Comment by quozl** [comment]
As `sprites.py` and `window_setup.py` are imported, and are used by Turtle Art, I'd like them to remain.

**Comment by yell0wfl4sh** [comment]
@quozl added both files.

**Pull request merged**